moveit_ros/moveit_servo/src/collision_monitor.cpp

```diff
@@ -0,0 +1,154 @@
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *******************************************************************************/
+/*
+ * Title      : collision_monitor.cpp
+ * Project    : moveit_servo
+ * Created    : 06/08/2023
+ * Author     : Brian O'Neil, Andy Zelenak, Blake Anderson, V Mohammed Ibrahim
+ */
+
+#include <moveit_servo/collision_monitor.hpp>
+#include <rclcpp/rclcpp.hpp>
+
+namespace
+{
+const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit_servo.servo");
+}
+
+namespace moveit_servo
+{
+
+CollisionMonitor::CollisionMonitor(const planning_scene_monitor::PlanningSceneMonitorPtr& planning_scene_monitor,
+                                   const servo::Params& servo_params, std::atomic<double>& collision_velocity_scale)
+  : servo_params_(servo_params)
+  , planning_scene_monitor_(planning_scene_monitor)
+  , collision_velocity_scale_(collision_velocity_scale)
+{
+}
+
+void CollisionMonitor::start()
+{
+  stop_requested_ = false;
+  if (!monitor_thread_.joinable())
+  {
+    monitor_thread_ = std::thread(&CollisionMonitor::checkCollisions, this);
+    RCLCPP_INFO_STREAM(LOGGER, "Collision monitor started");
+  }
+  else
+  {
+    RCLCPP_INFO_STREAM(LOGGER, "Collision monitor could not be started");
+  }
+}
+
+void CollisionMonitor::stop()
+{
+  stop_requested_ = true;
+  if (monitor_thread_.joinable())
+  {
+    monitor_thread_.join();
+  }
+  RCLCPP_INFO_STREAM(LOGGER, "Collision monitor stopped");
+}
+
+void CollisionMonitor::checkCollisions()
+{
+  rclcpp::WallRate rate(servo_params_.collision_check_rate);
+
+  bool approaching_self_collision, approaching_scene_collision;
+  double self_collision_threshold_delta, scene_collision_threshold_delta;
+  double self_collision_scale, scene_collision_scale;
+  double self_velocity_scale_coefficient{ -log(0.001) / servo_params_.self_collision_proximity_threshold };
+  double scene_velocity_scale_coefficient{ -log(0.001) / servo_params_.scene_collision_proximity_threshold };
+
+  while (rclcpp::ok() && !stop_requested_)
+  {
+    if (servo_params_.check_collisions)
```

Comment:
There's no `else` to this which simply sets the scaling to 1.0 -- should there be?

Reply:
Can be fixed by setting `collision_velocity_scale_` to 1.0 at the beginning of every loop :+1: